File: javascript/anagram/anagram.js

var comparableWord = function(word) {
  return word.toLowerCase().split('').sort().join()
}

var anagram = function(word) {
  var letters = comparableWord(word)
  var normalized = word.toLowerCase()

  var isAnagram = function(testWord) {
    return testWord.toLowerCase() != normalized &&
      comparableWord(testWord) == letters;
  }

  var matches = function(wordlist) {
    if (!Array.isArray(wordlist)) {
      wordlist = Array.prototype.slice.call(arguments);
    }
    return wordlist.filter(isAnagram)
  }
  return { matches: matches }
}

module.exports = anagram

File: scala/anagram/src/main/scala/Anagram.scala
class Anagram(val word: String) {
  val normalizedWord = normalize(word)

  def matches(others: Seq[String]): Seq[String] =
    others.filter(isAnagram(_))

  private def isAnagram(other: String): Boolean =
    other.toLowerCase != word.toLowerCase &&
      normalizedWord == normalize(other)

  private def normalize(word: String) =
    word.toLowerCase.sorted
}
